File: hacks/smbf/smbf.c
/*
 * ANSI C implementation of self-modifying brainfuck.
 * 
 * By Simon Howard <fraggle@removethisbit.gmail.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

unsigned char *tape;
unsigned int tape_size;
unsigned int tape_center;

void tape_init(void)
{
    tape_size = 1000;
    tape_center = tape_size / 2;
    tape = malloc(tape_size);
}

unsigned char tape_read(signed int index)
{
    index += tape_center;

    if (index < 0 || index >= tape_size) {
        return 0;
    } else {
        return tape[index];
    }
}

void tape_write(signed int index, unsigned char value)
{
    signed int index2;

    index2 = index + tape_center;

    if (index2 < 0 || index2 >= tape_size) {
        unsigned int new_tape_size;
        unsigned char *new_tape;
        unsigned int new_tape_center;

        /* Need to resize the tape to be larger */
        
        new_tape_size = tape_size;
        
        do {
            new_tape_size *= 2;
            new_tape_center = new_tape_size / 2;
            index2 = index + new_tape_center;
        } while (index2 < 0 || index2 >= new_tape_size);

        /* Now resized big enough */

        new_tape = malloc(new_tape_size);
        memset(new_tape, 0, new_tape_size);
        
        /* Copy old data */
        
        memcpy(new_tape + new_tape_center - tape_center, tape, tape_size);
        
        tape = new_tape;
        tape_size = new_tape_size;
        tape_center = new_tape_center;
    
        index2 = index + tape_center;
    }

    tape[index2] = value;
}

void interpret(signed int entrypoint)
{
    signed int pc;
    signed int ptr;
    unsigned char cmd;
    int loop_level;

    pc = entrypoint;
    ptr = 0;

    while (pc < 0) {

        cmd = tape_read(pc);

        switch(cmd) {
            case ',':
                tape_write(ptr, getchar());
                break;

            case '.':
                putchar(tape_read(ptr));
                break;

            case '<':
                ptr -= 1;
                break;

            case '>':
                ptr += 1;
                break;

            case '-':
                tape_write(ptr, tape_read(ptr) - 1);
                break;

            case '+':
                tape_write(ptr, tape_read(ptr) + 1);
                break;

            case '[':
                if (tape_read(ptr) == 0) {
                    loop_level = 1;

                    while (loop_level > 0) {
                        pc += 1;
                        if (tape_read(pc) == '[')
                            loop_level += 1;
                        else if (tape_read(pc) == ']')
                            loop_level -= 1;
                    }
                }
                break;

            case ']':
                /* Rewind to the loop start */

                loop_level = 1;

                while (loop_level > 0) {
                    pc -= 1;
                    if (tape_read(pc) == '[')
                        loop_level -= 1;
                    else if (tape_read(pc) == ']')
                        loop_level += 1;
                }

                pc -= 1;

                break;

            default:
                break;
        }

        pc += 1;
    }
}

signed int load_script(char *filename)
{
    FILE *fstream;
    signed int file_size;
    int i;
    
    fstream = fopen(filename, "rb");

    if (fstream == NULL) {
        fprintf(stderr, "Error loading '%s'\n", filename);
        exit(-1);
    }

    /* Find the file size */
     
    fseek(fstream, 0, SEEK_END);
    file_size = ftell(fstream);
    fseek(fstream, 0, SEEK_SET);

    /* Read in all data */

    for (i=0; i<file_size; ++i) {
        unsigned char value;

        fread(&value, 1, 1, fstream);

        tape_write(i - file_size, value);
    }

    return -file_size;
}

int main(int argc, char *argv[])
{
    signed int entrypoint;

    if (argc < 2) {
        printf("Usage: %s <script name>\n", argv[0]);
        exit(-1);
    }

    tape_init();
    entrypoint = load_script(argv[1]);
    interpret(entrypoint);

    return 0;
}

File: hacks/smbf/smbf.rb
#!/usr/bin/env ruby
#
# Self-modifying Brainfuck, by Simon Howard <fraggle@removethisbit.gmail.com>
#
# This behaves like normal brainfuck, except that the program is
# executed from the data tape.  The program text is inserted 
# immediately before the starting pointer position, ie. the following
# program prints the last character of the source file:
#
#        <.
#
# This allows for self-modifying code and introspection(reflection)
#

class Tape
    def initialize
        @data = "\0" * 1000
        @center = @data.length / 2
    end

    def [](index)
        @data[index+@center]
    end

    def []=(index, val)

        i = index + @center

        if i < 0 || i >= @data.length
            # resize the data array to be large enough

            new_size = @data.length

            loop do
                new_size *= 2
                test_index = index + (new_size / 2)
                if test_index >= 0 && test_index < new_size
                    # array is big enough now

                    break
                end
            end

            # generate the new array

            new_data = "\0" * new_size
            new_center = new_size / 2

            # copy old data into new array

            for j in 0...@data.length
                new_data[j-@center+new_center] = @data[j]
            end

            @data = new_data
            @center = new_center
        end

        @data[index+@center] = val & 0xff
    end
end

class Interpreter
    def initialize(data)
        @tape = Tape.new
        
        # copy the data into the tape

        for i in 0...data.length
            @tape[i-data.length] = data[i]
        end

        # program start point

        @entrypoint = -data.length
    end

    def char(c)
        sprintf("%c", c)
    end

    def call
        pc = @entrypoint
        ptr = 0

        while pc < 0
            case char(@tape[pc])
                when '.'
                    putc @tape[ptr]
                when ','
                    @tape[ptr] = STDIN.getc
                when '<'
                    ptr -= 1
                when '>'
                    ptr += 1
                when '-'
                    @tape[ptr] -= 1
                when '+'
                    @tape[ptr] += 1
                when '['
                    if @tape[ptr] == 0

                        # advance to end of loop

                        loop_level = 1

                        while loop_level > 0
                            pc += 1

                            loop_level += 1 if char(@tape[pc]) == '['
                            loop_level -= 1 if char(@tape[pc]) == ']'
                        end
                    end

                when ']'
                    # rewind to the start of the loop

                    loop_level = 1

                    while loop_level > 0
                        pc -= 1
                        loop_level -= 1 if char(@tape[pc]) == '['
                        loop_level += 1 if char(@tape[pc]) == ']'
                    end

                    pc -= 1
            end

            pc += 1
        end
    end
end

if ARGV.length <= 0
    puts "Usage: #{$0} <script filename>"
else
    File.open(ARGV[0]) do |file|
        data = file.readlines.join
        intr = Interpreter.new(data)
        intr.call
    end
end

